### PR TITLE
Baseline Comparison

### PR DESCRIPTION
**Pull request by miflower**
The purpose of this query is to perform a comparison between "known good" machines and suspected bad machines.  The following datasets are returned:
1. Alerts on the suspected bad machines (ignores known good machines, because...they're alerts, additional data has the triggered file)
2. Connected Networks (from MachineNetworkInfo table, additional data has full Connected Network details)
3. File Creations (disabled by default due to volume, enable at your own risk, additional data has initiating processes)
4. Image Loads (disabled by default due to volume, enable at your own risk, additional data has initiating processes)
5. Logon (derived from LogonEvents for the unique users logged on, additional data has logon types)
6. Network communication (grouped by 2nd level-domain, ie 'microsoft.com' in 'www.microsoft.com' and 'web.microsoft.com', additional data has the full list of URLs)
7. Process creation (additional data has the full paths of the files)
8. Powershell Commands (grouped by the cmdlet that was ran, additional data has the processes that ran the cmdlet)
9. Reigstry Events (disabled by default due to volume, grouped by the registry key, additional data has the value data)
10. Raw IP Connection Events (additional data has the initiating processes)

### CODE REVIEW

File: General queries/Baseline Comparison.txt [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    | where ComputerName in~ (InComputerName)
+    | summarize makeset(MachineId)
+};
+// Function to get network communications given a list of computer names and how far back to look
+let GetNetworkEvents=(InMachineId: dynamic, LeftEventTime: datetime) {
+    NetworkCommunicationEvents
+    | where "Network Communication" in (ReturnSets)
+    | where EventTime > LeftEventTime
+    | where MachineId in~ (InMachineId)
+    | summarize EventTime=max(EventTime), count() by RemoteUrl, MachineId
```
Comment (TomerAlpert) [inline General queries/Baseline Comparison.txt]: add:
| where isnotempty(RemoteUrl)

File: General queries/Baseline Comparison.txt [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    | extend UrlSplit=split(RemoteUrl, ".") // Split the levels of the URL
+    // If there is only one level (for an internal communication that uses your DNS search suffix), then only use that level
+    // Otherwise combine the top two levels and use those as the URLRoot
+    | extend UrlRoot=iff(UrlSplit[-2] == "", UrlSplit[0], strcat(tostring(UrlSplit[-2]), ".", tostring(UrlSplit[-1])))
+    | summarize EventTime=max(EventTime), Count=sum(count_), AdditionalData=makeset(RemoteUrl) by UrlRoot, MachineId
```
Comment (TomerAlpert) [inline General queries/Baseline Comparison.txt]: better limit the makeset, e.g. makeset(RemoteUrl, 5)

File: General queries/Baseline Comparison.txt [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+// Function to get process creates given a list of computer names and how far back to look
+let GetProcessCreates=(InMachineId: dynamic, LeftEventTime: datetime) {
+    ProcessCreationEvents
+    | where "Process Creation" in (ReturnSets)
+    | where EventTime > LeftEventTime
+    | where MachineId in~ (InMachineId)
+    // Replace known path for mpam files as they are dynamically named and likely to be unique on each machine
+    | extend FileName=iff(FolderPath matches regex @"([A-Z]:\\Windows\\ServiceProfiles\\NetworkService\\AppData\\Local\\Temp\\mpam-)[a-z0-9]{7,8}\.exe", "mpam-RANDOM.exe", FileName)
+    // Replace known path for AM delta patch files as they jump frequently and not likely to be exact on each machine
+    | extend FileName=iff(FolderPath matches regex @"([A-Z]:\\Windows\\SoftwareDistribution\\Download\\Install\\AM_Delta_Patch_)[0-9\.]+\.exe", "AM_Delta_Patch_Version.exe", FileName)
```
Comment (TomerAlpert) [inline General queries/Baseline Comparison.txt]: Making sure this is not DSRE specific

File: General queries/Baseline Comparison.txt [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+// Function to get powershell commands given a list of computer names and how far back to look
+let GetPSCommands=(InMachineId: dynamic, LeftEventTime: datetime) {
+    MiscEvents
+    | where "PowerShell Command" in (ReturnSets)
+    | where EventTime > LeftEventTime
+    | where MachineId in~ (InMachineId)
+    | where ActionType == 'PowerShellCommand'
+    // Remove two different signatures for scripts being executed which cause a lot of noise
+    // The first signature matches scripts generated as part of testing execution policy
+    // The second signature matches scripts generated by SCCM
```
Comment (TomerAlpert) [inline General queries/Baseline Comparison.txt]: maybe could add filtering on the initiatingprocessparentname, on initiating process commandline or change the regex to check also for the .ps1 suffix.

File: General queries/Baseline Comparison.txt [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+    | where not(AdditionalFields matches regex @"Script_[0-9a-f]{20}")
```
Comment (TomerAlpert) [inline General queries/Baseline Comparison.txt]: making sure this is not DSRE specific

File: General queries/Baseline Comparison.txt [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+    | extend iff(FileName matches regex @"[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\.ps1" or
+                 FileName matches regex @"[0-9a-z]{8}\.[0-9a-z]{3}\.ps1", "RANDOM.ps1", FileName)
+    | summarize EventTime=max(EventTime), FP_Set=makeset(FolderPath), count() by FileName, MachineId
+    | project EventTime, Entity=FileName, Count=count_, AdditionalData=tostring(FP_Set), MachineId, DataType="File Creation"
+};
+// Function to get logon events given a list of computer names and how far back to look
+let GetLogonEvents=(InMachineId: dynamic, LeftEventTime: datetime) {
+    LogonEvents
+    | where "Logon" in (ReturnSets)
+    | where EventTime > LeftEventTime
```
Comment (TomerAlpert) [inline General queries/Baseline Comparison.txt]: maybe you want to get a time-range (and use between), maybe you want different time-range per machine. If that's true, I suggest you use a data-table and join with it.

File: General queries/Baseline Comparison.txt [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+    | where MachineId in~ (InMachineId)
+    // Remove logons made by WDM or UMFD
+    | where AccountDomain !in ('font driver host', 'window manager')
+    // Replace qualys logons which cause a lot of noise
```
Comment (TomerAlpert) [inline General queries/Baseline Comparison.txt]: Seems DSRE specific to me

File: General queries/Baseline Comparison.txt [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+let Alerts=AlertEvents
+| where "Alert" in (ReturnSets)
+| where MachineId in (SuspectedBadHostMachineId)
+| summarize EventTime=max(EventTime), Count=count() by Title, MachineId, FileName, RemoteUrl
+| project EventTime, Entity=Title, Count, AdditionalData=coalesce(FileName, RemoteUrl), DataType="Alert";
+// String everything together
+union NetworkDelta, ProcessDelta, PSDelta, FileDelta, Alerts, LogonDelta, RegistryDelta,
+      ConnectedNetworkDelta, ImageLoadDelta, RawIPCommunicationDelta
+// Join back against the machine info so the Computer Names can be reassociated
+| join kind=leftouter (
```
Comment (TomerAlpert) [inline General queries/Baseline Comparison.txt]: probably better to select ComputerName in the first place

File: General queries/Baseline Comparison.txt [hunk 9]
```diff
@@ -0,0 +1,239 @@
+// Baseline Comparison
+// Author: miflower
+// The purpose of this query is to perform a comparison between "known good" machines and suspected bad machines
+// The original concept for this query was born due to reapplying the same 'whitelist' filters over and over
+// It brings deltas between a baseline and another machine quickly to the analyst's view
+// This query supports multiple suspected bad machines and multiple "known good" machines
+// It also supports providing a timeframe for how far back in time to build a baseline as well as how far back in time to evaluate the suspected bad machines
+// Each of the links provided by machineid/computername will go to the most recent entry for whatever entity is listed
+// Average results for the pre-defined settings below with a single good host and a single bad host:
```
Comment (TomerAlpert) [inline General queries/Baseline Comparison.txt]: These stats are DSRE specific (especially the compute time).
Either remove it or comment that it's the compute time on a huge tenant.

File: General queries/Baseline Comparison.txt [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+      ConnectedNetworkDelta, ImageLoadDelta, RawIPCommunicationDelta
+// Join back against the machine info so the Computer Names can be reassociated
+| join kind=leftouter (
+    MachineInfo
+    | where MachineId in (GoodHostMachineId, SuspectedBadHostMachineId)
+    | distinct MachineId, ComputerName
+) on MachineId
+| project-away MachineId1
+// and sort by Machine, DataType, Entity
+| order by MachineId asc, DataType asc, Entity asc
```
Comment (TomerAlpert) [inline General queries/Baseline Comparison.txt]: Maybe add something like:
| summarize arg_max(EventTime, AdditionalData, MachineId), BadMachinesCount=dcount(MachineId) by Entity, DataType
| where BadMachinesCount > 1

File: General queries/Baseline Comparison.txt (hunk 3 above)
Comment (miflower) [reply General queries/Baseline Comparison.txt]: AM and MPAM are both not DSRE specific

File: General queries/Baseline Comparison.txt (hunk 6 above)
Comment (miflower) [reply General queries/Baseline Comparison.txt]: I discussed this with my time for quite awhile actually.  There are lots of opportunities for customizing the filtering being done.  In the end we decided it is best for most analysts to have the few options at the beginning.  If you want to add those types of modifications in, each analyst should know how to do that within AH themselves.  We really tried to keep it as simple and easy to use for all analysts.

File: General queries/Baseline Comparison.txt (hunk 4 above)
Comment (miflower) [reply General queries/Baseline Comparison.txt]: Added in latest commits

File: General queries/Baseline Comparison.txt (hunk 1 above)
Comment (miflower) [reply General queries/Baseline Comparison.txt]: Updated in latest commit.

File: General queries/Baseline Comparison.txt (hunk 2 above)
Comment (miflower) [reply General queries/Baseline Comparison.txt]: Updated in latest commit.

File: General queries/Baseline Comparison.txt (hunk 5 above)
Comment (miflower) [reply General queries/Baseline Comparison.txt]: Not DSRE specific.

File: General queries/Baseline Comparison.txt (hunk 7 above)
Comment (miflower) [reply General queries/Baseline Comparison.txt]: Is DSRE-specific, removed in latest commit.

File: General queries/Baseline Comparison.txt (hunk 8 above)
Comment (miflower) [reply General queries/Baseline Comparison.txt]: Updated in latest commit.

File: General queries/Baseline Comparison.txt (hunk 9 above)
Comment (miflower) [reply General queries/Baseline Comparison.txt]: Added comment ;)

File: General queries/Baseline Comparison.txt (hunk 10 above)
Comment (miflower) [reply General queries/Baseline Comparison.txt]: Really liked this idea and actually took it in another direction.
Seeing the prevalence of the deviation is definitely a useful data point to have but limiting results based off each tended to really reduce the value of the query.  We really wanted to keep the clickable links to the timeline for further investigation.  As such, I added in logic to calculate the BadMachinesCount, then added it to each row.

**Pull request merged**